zkvm/src/constraints.rs

```diff
@@ -44,6 +45,43 @@ pub struct CommitmentWitness {
     blinding: Scalar,
 }
 
+impl Constraint {
+    pub fn verify<CS: r1cs::ConstraintSystem>(self, cs: &mut CS) -> Result<(), VMError> {
+        cs.specify_randomized_constraints(move |cs| {
+            // Flatten the constraint into one expression
+            // Note: cloning because we can't move out of captured variable in an `Fn` closure,
+            // and `Box<FnOnce>` is not fully supported yet. (We can update when that happens).
+            // Cf. https://github.com/dalek-cryptography/bulletproofs/issues/244
+            let expr = self.clone().flatten(cs);
+
+            // Add the resulting expression to the constraint system
+            cs.constrain(expr.to_r1cs_lc());
+
+            Ok(())
+        })
+        .map_err(|e| VMError::R1CSError(e))
+    }
+
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS) -> Expression {
+        match self {
+            Constraint::Eq(expr1, expr2) => expr1 - expr2,
+            Constraint::And(c1, c2) => {
+                let a = c1.flatten(cs);
+                let b = c2.flatten(cs);
+                // output expression: a + z * b
+                let z = Expression::constant(cs.challenge_scalar(b"verify challenge"));
+                a + b.multiply(z, cs)
```

Comment:
If these are r1cs::LinearCombinations, we can simply do 

```rust
let z = cs.challenge_scalar(b"ZkVM.verify.and-challenge");
a + b.multiply(z, cs)
```

BTW, we probably need to update the spec with the label `"ZkVM.verify.and-challenge"` because it affects the resulting proof.

Reply:
👍